packages/xstate-wallet/src/workflows/close-ledger-and-withdraw.ts

```diff
@@ -1,87 +1,173 @@
-  getFinalState: (context: WorkflowContext, event: WorkflowEvent) => Promise<SupportState.Init>;
-  supportState: StateMachine<any, any, any>;
-  submitWithdrawTransaction: (context: WorkflowContext, event: WorkflowEvent) => Promise<void>;
-  createObjective: (context: WorkflowContext, event: any) => Promise<Objective>;
+
+enum Services {
+  constructFinalState = 'constructFinalState',
+  supportState = 'supportState',
+  submitWithdrawTransaction = 'submitWithdrawTransaction',
+  createObjective = 'createObjective',
+  observeFundsWithdrawal = 'observeFundsWithdrawal',
+  fetchBudget = 'fetchBudget',
+  clearBudget = 'clearBudget'
 }
 
-export const config: StateNodeConfig<WorkflowContext, any, any> = {
+export type WorkflowActions = CommonWorkflowActions &
+  Record<
+    Actions,
+    ActionFunction<WorkflowContext, WorkflowEvent> | AssignAction<WorkflowContext, WorkflowEvent>
+  >;
+export type WorkflowServices = Record<Services, ServiceConfig<WorkflowContext>>;
+
+export const config: MachineConfig<WorkflowContext, any, WorkflowEvent> = {
   id: 'close-and-withdraw',
-  initial: 'waitForUserApproval',
+
+  initial: 'fetchBudget',
   states: {
+    fetchBudget: {
+      invoke: {
+        src: Services.fetchBudget,
+        onDone: {target: 'waitForUserApproval', actions: [Actions.assignBudget]}
+      }
+    },
     waitForUserApproval: {
-      entry: ['displayUi'],
-
+      entry: [CommonActions.displayUI],
       on: {
-        USER_APPROVES_CLOSE: [
-          {cond: 'doesChannelIdExist', target: 'closeLedger'},
-          'createObjective'
-        ],
-        USER_REJECTS_CLOSE: {target: 'failure'}
+        USER_APPROVES_CLOSE: {target: 'createObjective'},
+        USER_REJECTS_CLOSE: {target: 'userDeclinedFailure'}
       }
     },
 
     createObjective: {
       invoke: {
-        src: 'createObjective',
-        onDone: {target: 'closeLedger', actions: ['assignLedgerId']}
+        src: Services.createObjective,
+        onDone: {target: 'closeLedger', actions: [Actions.assignLedgerId]}
       }
     },
-    closeLedger: getDataAndInvoke({src: 'getFinalState'}, {src: 'supportState'}, 'withdraw'),
+    closeLedger: getDataAndInvoke<LedgerExists>(
+      {src: Services.constructFinalState},
+      {src: Services.supportState},
+      'withdraw'
+    ) as any,
     withdraw: {
-      entry: ['clearBudget'],
-      invoke: {src: 'submitWithdrawTransaction', onDone: 'done', onError: 'failure'}
+      initial: 'submitTransaction',
+      on: {
+        FUNDS_WITHDRAWN: 'clearBudget'
+      },
+      invoke: {
+        id: 'observeChain',
+        src: Services.observeFundsWithdrawal
+      },
+      states: {
+        submitTransaction: {
+          invoke: {
+            id: 'submitTransaction',
+            src: Services.submitWithdrawTransaction,
+            onDone: {
+              target: 'waitMining',
+              actions: [Actions.setTransactionId]
```

Comment:
Not expecting a change in this PR, but I'm strongly in favour of removing away from defining config in terms of strings, and move towards putting the functions in directly. This opens the path to better typesafety (you see it immediately with actions on named events, which are checked against the actual action type they will be passed), and also to better code discovery, by being able to directly find the code that will execute at that point, instead of having to trace through to where it is plugged together.

Reply:
Does defining the action functions inline work with the xstate visualizer? I noticed that currently approve budget and fund is not generating a config for the visualizer.

Reply:
Is `xstate-devtools` the visualizer or the chrome extension?